topdown/object.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+func builtinObjectUnion(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	objA, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	objB, err := builtins.ObjectOperand(operands[1].Value, 2)
+	if err != nil {
+		return err
+	}
+
+	r := mergeWithOverwrite(objA, objB)
+
+	return iter(ast.NewTerm(r))
+}
+
+func builtinObjectRemove(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Expect an object and an array/set of keys
+	obj, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Build a set of keys to remove
+	keysToRemove := getObjectKeysParam(operands[1].Value)
+
+	r := ast.NewObject()
+	obj.Foreach(func(key *ast.Term, value *ast.Term) {
+		if !keysToRemove.Contains(key) {
+			r.Insert(key, value)
+		}
+	})
+
+	return iter(ast.NewTerm(r))
+}
+
+func builtinObjectFilter(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Expect an object and an array/set of keys
+	obj, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Build a new object from the supplied filter keys
+	keys := getObjectKeysParam(operands[1].Value)
+	filterObj := ast.NewObject()
+	keys.Foreach(func(key *ast.Term) {
+		filterObj.Insert(key, ast.NullTerm())
+	})
+
+	// Actually do the filtering
+	r, err := obj.Filter(filterObj)
+	if err != nil {
+		return err
+	}
+
+	return iter(ast.NewTerm(r))
+}
+
+func builtinObjectGet(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	object, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	if ret := object.Get(operands[1]); ret != nil {
+		return iter(ret)
+	}
+
+	return iter(operands[2])
+}
+
+
+// getObjectKeysParam returns a set of key values
+// from a supplied ast string, array, or set value
```

Comment:
This comment appears to be out of date. It does not consider strings.

Reply:
Since we're returning errors if the user supplies a non-object value for the other operands, I think this function should return an error if the keys are not an array/set/object. Otherwise these built-ins will return the input object as-is.

Reply:
Done